src/YarpUtilities/src/VectorsCollectionServer.cpp

```diff
@@ -0,0 +1,176 @@
+    {
+        log()->error("{} Unable to attach the metadata port.", logPrefix);
+        return false;
+    }
+
+    return true;
+}
+
+bool VectorsCollectionServer::populateMetadata(const std::string& key,
+                                               const std::vector<std::string>& metadata)
+{
+    // check if the metadata has been already populated
+    if (m_pimpl->isMetadataFinalized)
+    {
+        log()->error("[VectorsCollectionServer::populateMetadata] The metadata has been already "
+                     "populated.");
+        return false;
+    }
+
+    // check if the key has been already used
+    if (m_pimpl->setOfKeys.find(key) != m_pimpl->setOfKeys.end())
+    {
+        log()->error("[VectorsCollectionServer::populateMetadata] The key {} has been already "
+                     "used.",
+                     key);
+        return false;
+    }
+
+    // populate the metadata
+    m_pimpl->metadata.vectors[key] = metadata;
+
+    // add the key to the set of keys
+    m_pimpl->setOfKeys.insert(key);
+
+    return true;
+}
+
+bool VectorsCollectionServer::finalizeMetadata()
+{
+    // check if the metadata has been already finalized
+    if (m_pimpl->isMetadataFinalized)
+    {
+        log()->error("[VectorsCollectionServer::finalizeMetadata] The metadata has been already "
+                     "finalized.");
+        return false;
+    }
+
+    // check if the metadata is empty
+    if (m_pimpl->metadata.vectors.empty())
+    {
+        log()->error("[VectorsCollectionServer::finalizeMetadata] The metadata is empty.");
+        return false;
+    }
+
+    // set the metadata as finalized
+    m_pimpl->isMetadataFinalized = true;
+
+    return true;
+}
+
+bool VectorsCollectionServer::populateData(const std::string& key,
+                                           const iDynTree::Span<const double>& data)
+{
+    constexpr auto logPrefix = "[VectorsCollectionServer::setData]";
+
+    // check if the metadata has been finalized
+    if (!m_pimpl->isMetadataFinalized)
+    {
+        log()->error("{} The metadata has not been finalized.", logPrefix);
+        return false;
+    }
+
+    // check if the key exists
+    if (m_pimpl->setOfKeys.find(key) == m_pimpl->setOfKeys.end())
+    {
+        log()->error("{} The key {} does not exist.", logPrefix, key);
+        return false;
+    }
+
+    // prepare the data
+    BipedalLocomotion::YarpUtilities::VectorsCollection& collection = m_pimpl->port.prepare();
+    collection.vectors[key].assign(data.begin(), data.end());
+
+    return true;
+}
+
+void VectorsCollectionServer::write(bool forceStrict /*= false */)
+{
+    m_pimpl->port.write(forceStrict);
+}
+
+void VectorsCollectionServer::clear()
+{
+    m_pimpl->port.prepare().vectors.clear();
+}
+
+bool VectorsCollectionServer::areMetadataReady()
+{
+    return m_pimpl->isMetadataFinalized;
+}
+
+VectorsCollectionMetadata VectorsCollectionServer::getMetadata()
+{
+    return m_pimpl->metadata;
```

Comment:
I think we need to check whether the metadata is finalized first

Reply:
This is defined in the thrift. I don't know if it is possible to return two values (probably with a pair).
What do you think?

Reply:
In case you could just return an empty ``VectorsCollectionMetadata``. I guess it is better than a malformed one. This would also allow avoiding concurrency issues with clients that try to access the metadata before they are finalized.

Reply:
Ok perfect! I will go for an empty one :)